server/mdm/microsoft/wstep.go

```diff
@@ -0,0 +1,223 @@
+package microsoft_mdm
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/micromdm/nanomdm/cryptoutil"
+	"go.mozilla.org/pkcs7"
+)
+
+// TODO: Replace with imports from Marcos' PR
+const (
+	DocProvisioningAppProviderID = "FleetDM"
+	CertRenewalPeriodInSecs      = "15552000"
+)
+
+// TODO: Replace with imports from Marcos' PR
+type BinSecTokenType int
+
+// TODO: Replace with imports from Marcos' PR
+const (
+	MDETokenPKCS7 BinSecTokenType = iota
+	MDETokenPKCS10
+	MDETokenPKCSInvalid
+)
+
+// WSTEPDepot implements certificate management tasks associated with MS-WSTEP messages in the MS-MDE2
+// protocol.
+//
+// Note: The current implementation does not persist certificates it issues; however, it could
+// be extended to do so in the future. In that case, the depot would need to be initialized with a
+// datastore connection and extended to include datastore operations (similar to the SCEP depot).
+type WSTEPDepot struct {
+	// identityCert holds the identity certificate of the depot.
+	identityCert *x509.Certificate
+	// identityPrivateKey holds the private key of the depot.
+	identityPrivateKey *rsa.PrivateKey
+	// identityFingerprint holds the hex-encoded, sha1 fingerprint of the identity certificate.
+	IdentityFingerprint *string
+	// maxSerialNumber holds the maximum serial number. The maximum value a serial number can have
+	// is 2^160. However, this could be limited further if required.
+	maxSerialNumber *big.Int
+}
+
+// CertDepot is an interface for certificate management tasks associated with Microsoft MDM (e.g.,
+// signing CSRs). It is implemented by *WSTEPDepot.
+type CertDepot interface {
+	SignClientCSR(subject string, clientCSR *x509.CertificateRequest) ([]byte, string, error)
+}
+
+// newWSTEPDepot creates and returns a *WSTEPDepot.
+func NewWSTEPDepot(certPEM []byte, privKeyPEM []byte) (*WSTEPDepot, error) {
+	crt, err := cryptoutil.DecodePEMCertificate(certPEM)
+	if err != nil {
+		return nil, err
+	}
+	key, err := decodeRSAKeyFromPEM(privKeyPEM)
+	if err != nil {
+		return nil, err
+	}
+	fp := CertFingerprintHexStr(crt)
+
+	return &WSTEPDepot{
+		identityCert:        crt,
+		identityPrivateKey:  key,
+		IdentityFingerprint: &fp,
+		maxSerialNumber:     new(big.Int).Lsh(big.NewInt(1), 128), // 2^12,
+	}, nil
+}
+
+func decodeRSAKeyFromPEM(key []byte) (*rsa.PrivateKey, error) {
+	block, _ := pem.Decode(key)
+	if block == nil {
+		return nil, errors.New("failed to decode PEM private key")
+	}
+	if block.Type != "RSA PRIVATE KEY" {
+		return nil, errors.New("PEM type is not RSA PRIVATE KEY")
+	}
+	return x509.ParsePKCS1PrivateKey(block.Bytes)
+}
+
+func CertFingerprintHexStr(cert *x509.Certificate) string {
+	fingerprint := sha1.Sum(cert.Raw)
```

Comment:
Same as before, I've found a way to use SHA256WithRSA certificate generation, but Windows Certificate Store requires passing the certificate thumbprint (same as SHA1 fingerprint details - see [here](https://security.stackexchange.com/questions/14330/what-is-the-actual-value-of-a-certificate-fingerprint)). So we still have to use sha1 for this. This does not affect the security of the certificate (see [here](https://www.thesslstore.com/blog/ssl-certificate-still-sha-1-thumbprint/)) and it is mostly used as a way to locate the cert on the cert store.

Having said that, I'll suggest to add //nolint:gosec here to avoid the lint warning and leave the code as is - I'll add SHA256WithRSA support on my PR